optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/DroolsAccumulateContext.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+
+/**
+ * Each context is uniquely identified by its ID.
+ * This is necessary so that the Drools accumulate function can properly undo in
+ * {@link DroolsUniAccumulateFunctionBridge} and its Bi, Tri, ... alternatives.
+ * @param <ResultContainer_> The same type from {@link UniConstraintCollector} and its Bi, Tri, ... alternatives.
+ */
+public final class DroolsAccumulateContext<ResultContainer_> implements Serializable {
+
+    private static final AtomicLong CONTEXT_COUNTER = new AtomicLong();
```

Comment:
Can we get rid of making this static?

In multithreaded-solving, multiple score directors (each with it's own CS session) will be created in parallels. Normally they don't feel each other: each one is used on it's own thread, so they are single-threaded and don't share anything.
But this static CONTEXT_COUNTER will be shared. So the calls to it will be in a different order between runs, affecting the DroolsAccumulateContext's id (which is a pain for debugging and reproducible debugging). 

Furthermore, when we scale to 100 move threads, this means a 100 move threads are concurrently hitting that CONTEXT_COUNTER for every accumulate during session creation I presume. That's probably not the end of the world, as that congestion will last only a few milliseconds, but it's still nice to avoid.

Reply:
I see your point. That said, it is `static` because the counter needs to be incremented. I'm not sure how to do this without having a single source of truth *somewhere.*

Reply:
Well, the IDs needn't be in sequence. They may also be random. But I'm not sure that this is better.